test/embind/test_val_integers.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright 2023 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+#include <emscripten.h>
+#include <emscripten/val.h>
+#include <iostream>
+#include <limits>
+
+using namespace emscripten;
+
+bool check_num_on_js_side(std::string value_s) {
+  std::string code = std::string("num === ") + value_s + " || console.error(`\t\tFailed: expected " + value_s + ", got ${num}`)";
+  return emscripten_run_script_int(code.c_str()) == 1;
+}
+
+template <typename T>
+bool check_num_on_cpp_side(T expected) {
+  T value = emscripten::val::global("num").as<T>();
+  if (value != expected) {
+    std::cerr << "\t\tFailed: expected " << expected << ", got " << value << std::endl;
+    return false;
+  }
+  return true;
+}
+
+template <typename T>
+bool test_num(const char *type_name, T value) {
+  std::string value_s = std::to_string(value);
+  // 64-bit integer must be represented as a BigInt which uses `n` suffix
+  if (sizeof(T) > 4) {
+    value_s.push_back('n');
+  }
+  bool ok = true;
+  std::cout << "Testing (" << type_name << ") " << value_s << std::endl;
+  // Test C++ to JS conversion
+  std::cout << "\tC++ to JS via assignment" << std::endl;
+  emscripten::val::global().set("num", value);
+  ok &= check_num_on_js_side(value_s);
+  // Test C++ to JS conversion via function call.
+  // Calls use their own argument serialization, so it's best to ensure that works too.
+  std::cout << "\tC++ to JS via call" << std::endl;
+  emscripten::val::global("setNum")(value);
+  ok &= check_num_on_js_side(value_s);
+  // Test JS to C++ conversion
+  std::cout << "\tJS to C++" << std::endl;
+  ok &= check_num_on_cpp_side(value);
+  return ok;
+}
+
+template <typename T>
+bool test_num_limits(const char *type_name) {
+  // note: using `&` instead of `&&` to ensure both tests are run
+  return test_num<T>(type_name, std::numeric_limits<T>::min()) &
+         test_num<T>(type_name, std::numeric_limits<T>::max());
+}
+
+#define TEST_NUM_LIMITS(T) ok &= test_num_limits<T>(#T)
+
+int main() {
+  bool ok = true;
```

Comment:
> It means test will exit on first failure but I've found that that is almost always what I want anyway

That's what I _almost always_ want too, but I believe in this case an exception was warranted: there are various failure edge cases and the goal was to uncover all of them in a flat list. Otherwise we wouldn't know what other bugs are there until we fix the first one and the next one is uncovered and so on.

When all those bugs are fixed, we can revert this to simpler version if that's desirable, but for now it was more of a "show all types that are broken" exploration.

Reply:
(worth noting that it still doesn't fully realise that goal because at some point test crashes with JS exception after regular value mismatches, so there are probably more things broken after that line)